compiler/src/dotty/tools/dotc/core/TyperState.scala

```diff
@@ -229,7 +229,10 @@ class TyperState() {
         constraint.contains(tl) || other.isRemovable(tl) || {
           val tvars = tl.paramRefs.map(other.typeVarOfParam(_)).collect { case tv: TypeVar => tv }
           if this.isCommittable then
-            tvars.foreach(tvar => if !tvar.inst.exists && !isOwnedAnywhere(this, tvar) then includeVar(tvar))
+            tvars.foreach(tvar =>
+              if !tvar.inst.exists then
+                if !isOwnedAnywhere(this, tvar) then includeVar(tvar)
+                if other.isHard(tvar) then constraint = constraint.withHard(tvar))
```

Comment:
This code path is only executed if `constraint.contains(tl)` is false above, but if the same type lambda appears in both constraints, they could still have different sets of hardened tvars.  So I think this logic needs to be moved below inside the argument to `constraint.uninstVars.forall`.

Reply:
Ah yes, well spotted.